c/cert/src/rules/EXP32-C/DoNotAccessVolatileObjectWithNonVolatileReference.ql

```diff
@@ -0,0 +1,83 @@
+/**
+ * @id c/cert/do-not-access-volatile-object-with-non-volatile-reference
+ * @name EXP32-C: Do not access a volatile object through a nonvolatile reference
+ * @description If an an object defined with a volatile-qualified type is referred to with an lvalue
+ *              of a non-volatile-qualified type, the behavior is undefined.
+ * @kind problem
+ * @precision high
+ * @problem.severity error
+ * @tags external/cert/id/exp32-c
+ *       correctness
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+import semmle.code.cpp.controlflow.Dereferenced
+
+/**
+ * An expression involving volatile-qualified types that results in undefined behavior.
+ */
+abstract class UndefinedVolatilePointerExpr extends Expr {
+  /**
+   * Gets a descriptive string describing the type of expression and undefined behavior.
+   */
+  abstract string getMessage();
+}
+
+/**
+ * Gets the depth of a pointer's base type's volatile qualifier
+ */
+int getAVolatileDepth(PointerType pt) {
```

Comment:
I see the issue with `PointerType` and have changed it to `DerivedType`, but since this predicate only holds for volatile base types, would there actually be a false-negative case? `DerivedType` should also include types with volatile specifiers. The `volatile int* volatile` with a volatile specifier on the pointer rather than its base type seems to not impact the rule, which focuses on volatile specifiers on base types as far as I understand.